### PR TITLE
Fix "except Exception" in limited api

### PR DESCRIPTION
**Pull request by da-woods**
Partial fix for #5697

Change cname for builtin types to be a pointer rather than a value. This means that we can avoid defererencing BaseException to transform it into a value (which doesn't work on the limited api, because its an opaque struct)

### CODE REVIEW

**Comment by da-woods** [comment]
Unfortunately this doesn't look like it;s going to work in this form

File: Cython/Compiler/TypeSlots.py [hunk 1]
```diff
@@ -717,7 +717,7 @@ class DictOffsetSlot(SlotDescriptor):
     def slot_code(self, scope):
         dict_entry = scope.lookup_here("__dict__") if not scope.is_closure_class_scope else None
         if dict_entry and dict_entry.is_variable:
-            if getattr(dict_entry.type, 'cname', None) != 'PyDict_Type':
+            if getattr(dict_entry.type, 'typeptr_cname', None) not in ['&PyDict_Type', '(&PyDict_Type)']:
                 error(dict_entry.pos, "__dict__ slot must be of type 'dict'")
                 return "0"
```
Comment (scoder) [inline Cython/Compiler/TypeSlots.py:722]: Not the fault of this PR, but it seems very late do detect this case here, during code generation.
But shouldn't we just check `is Builtin.dict_type` here? Why look at the type pointer?

**Review by scoder — COMMENTED** [review]
Looks good, but refactorings like this always feel risky for a point release. Should we do this in Cython 3.1?

**Comment by da-woods** [comment]
> Looks good, but refactorings like this always feel risky for a point release. Should we do this in Cython 3.1?

Probably sensible. I don't think there's an alternative way of way of the issue but there's plenty of other things broken in the limited API anyway.

> But shouldn't we just check is Builtin.dict_type here? Why look at the type pointer?

I'd assumed it was too do with letting you also use a `cdef extern` dict redeclaration. But I don't really see why that should be necessary.

**Comment by da-woods** [comment]
In the absence of any objections I'm going to treat this as "approved for 3.1" and merge it in fairly shortly in the hope that any regressions will be caught early.

**Pull request merged**